### PR TITLE
Refactor and add new tests to hugepages e2e tests

### PR DESCRIPTION
**Pull request by cynepco3hahue**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Add e2e tests to cover usage of multiple hugepages with different
page sizes under the same pod.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
It should be merged together with new prow jobs - https://github.com/kubernetes/test-infra/pull/17634

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [Usage]: https://github.com/kubernetes/enhancements/blob/8137e8370a55b789f17fbe5e674a690063a5b735/keps/sig-node/20190129-hugepages.md#pod-specification
```

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

### CODE REVIEW

**Comment by cynepco3hahue** [comment]
I unsure what kind should I set.

**Comment by derekwaynecarr** [comment]
/approve
/lgtm

This is an improvement over the prior test by testing more than one page size.  I think we could restructure this test in the future to detect all huge page sizes supported, and attempt to use each page size discovered.

**Comment by cynepco3hahue** [comment]
> This is an improvement over the prior test by testing more than one page size. I think we could restructure this test in the future to detect all huge page sizes supported, and attempt to use each page size discovered.

I totally agree it can be useful to test on top of different architectures.

File: test/e2e_node/hugepages_test.go [hunk 1]
```diff
@@ -90,145 +116,309 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
...
+		for hugepagesResource, count := range hugepages {
+			size := resourceToSize[hugepagesResource]
+			ginkgo.By("Verifying hugepages are supported")
+			if !isHugePageSupported(size) {
+				e2eskipper.Skipf("skipping test because hugepages of size %d not supported", size)
 				return
 			}
-			ginkgo.By("configuring the host to reserve a number of pre-allocated hugepages")
+
+			ginkgo.By("Configuring the host to reserve a number of pre-allocated hugepages")
```
Comment (bsdnet) [inline test/e2e_node/hugepages_test.go:207]: Same here if we can include size and count

File: test/e2e_node/hugepages_test.go [hunk 2]
```diff
@@ -33,30 +36,53 @@ import (
...
+	cgroupFsName := ""
 	if framework.TestContext.KubeletConfig.CgroupDriver == "systemd" {
 		cgroupFsName = cgroupName.ToSystemd()
 	} else {
 		cgroupFsName = cgroupName.ToCgroupfs()
 	}
 
-	command := ""
+	hugetlbLimitFile := ""
 	// this command takes the expected value and compares it against the actual value for the pod cgroup hugetlb.2MB.<LIMIT>
```
Comment (bsdnet) [inline test/e2e_node/hugepages_test.go:78]: hugetlb.2MB ==> e.g hugetlb.2MB to illustrate the format ?

Comment (cynepco3hahue) [reply test/e2e_node/hugepages_test.go:78]: Can you please clarify, I unsure that I understand the comment?

File: test/e2e_node/hugepages_test.go [hunk 3]
```diff
@@ -90,145 +116,309 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
...
-		err := exec.Command("/bin/sh", "-c", "echo 1 > /proc/sys/vm/compact_memory").Run()
-		if err != nil {
+		if err := exec.Command("/bin/sh", "-c", "echo 1 > /proc/sys/vm/compact_memory").Run(); err != nil {
 			return err
 		}
 	}
-	err := exec.Command("/bin/sh", "-c", "echo 5 > /proc/sys/vm/nr_hugepages").Run()
-	if err != nil {
+
+	command := fmt.Sprintf("echo %d > %s-%dkB/%s", hugepagesCount, hugepagesDirPrefix, hugepagesSize, hugepagesCapacityFile)
```
[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply test/e2e_node/hugepages_test.go:130]: done

File: test/e2e_node/hugepages_test.go [hunk 4]
```diff
@@ -90,145 +116,309 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
...
+	if err := exec.Command("/bin/sh", "-c", command).Run(); err != nil {
 		return err
 	}
-	outData, err := exec.Command("/bin/sh", "-c", "cat /proc/meminfo | grep 'HugePages_Total' | awk '{print $2}'").Output()
+
+	command = fmt.Sprintf("cat %s-%dkB/%s", hugepagesDirPrefix, hugepagesSize, hugepagesCapacityFile)
```
[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply test/e2e_node/hugepages_test.go:135]: done

File: test/e2e_node/hugepages_test.go [hunk 5]
```diff
@@ -90,145 +116,309 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
...
 
-	ginkgo.Context("With config updated with hugepages feature enabled", func() {
-		ginkgo.BeforeEach(func() {
-			ginkgo.By("verifying hugepages are supported")
-			if !isHugePageSupported() {
-				e2eskipper.Skipf("skipping test because hugepages are not supported")
+	setHugepages := func() {
+		for hugepagesResource, count := range hugepages {
+			size := resourceToSize[hugepagesResource]
+			ginkgo.By("Verifying hugepages are supported")
```
[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply test/e2e_node/hugepages_test.go:201]: done

File: test/e2e_node/hugepages_test.go (hunk 1 above)
Comment (cynepco3hahue) [reply test/e2e_node/hugepages_test.go:207]: done

File: test/e2e_node/hugepages_test.go [hunk 6]
```diff
@@ -90,145 +116,309 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
...
-	return amount
+	return fmt.Errorf("expected hugepages %v, but found %v", hugepagesCount, numHugePages)
 }
 
-// amountOfResourceAsString returns the amount of resourceName advertised by a node
-func amountOfResourceAsString(node *v1.Node, resourceName string) string {
-	val, ok := node.Status.Capacity[v1.ResourceName(resourceName)]
-	if !ok {
-		return ""
+// isHugePageSupported returns true if hugepages of the specified size is available on the host
```
[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply test/e2e_node/hugepages_test.go:154]: done

**Comment by cynepco3hahue** [comment]
@bart0sh Can you please review again?

**Review by odinuge — APPROVED** [review]
Thanks for the cleanup! As other said, it would also be nice with support for other sizes, but this is a great start. Ref. https://github.com/kubernetes/kubernetes/pull/78495 arm64 support 4 different sizes. ~As far as I remember the CI infra on gce only support 2MiB huge pages, but that would only skip the 1GiB pages, so that should be no issue~.

I haven't run the tests locally, but the changes looks good to me. Since this is already approved, ill add a hold so it isn't merged right away. In case some other reviewer or you @cynepco3hahue would like to merge it, feel free to cancel the hold. :smile: 

/lgtm
/kind cleanup
/hold

**Comment by cynepco3hahue** [comment]
@odinuge Thanks for the review, in case, if you want to test hugepages under the CI, we have now a separate job for it - `pull-kubernetes-node-kubelet-serial-hugepages, currently it supports 2Mi and 1Gi hugepages.

**Comment by odinuge** [comment]
> You want to test hugepages under the CI, we have now a separate job for it - `pull-kubernetes-node-kubelet-serial-hugepages, currently it supports 2Mi and 1Gi hugepages.

Yes, saw that just after leaving the review! Fantastic!

Here is the testgridlink for those who are looking for it: https://testgrid.k8s.io/sig-node-kubelet#kubelet-serial-gce-e2e-hugepages :smile:

**Review by odinuge — APPROVED** [review]
Another lgtm from me, as this looks good! :smile: Agree that it is smart to move it to a NodeSpecialFeature so we don't "help" ci-kubernetes-node-kubelet-serial to flake due to memory pressure.

I think this is good to go, so no hold this time.

/lgtm

**Pull request merged**